examples/5.MIS/main.jl

```diff
@@ -350,3 +350,280 @@ Bloqade.plot!(ax, Δ_final)
+# Special circuitry is added in the scribing corridors that enables adjacent chips to be connected. 
+# Then, after the wafer is tested, adjacent functional chips are connected into a group and the entire group is used as a single enlarged chip." 
+# In this way, even if a smaller chips becomes defective, the adjacent chips can be rewired so that they can still be used in the larger chip. 
+
+# A natural problem that arises from this setup is the _square packing_ problem. 
+# [D. Hochbaum and W. Maass](https://www.cs.du.edu/~snarayan/sada/research/docs/p130-hochbaum.pdf) summarizes an example problem: 
+# "64K RAM chips, some of which may be defective, are available on a rectilinear grid placed on a silicon wafer. 
+# 2 x 2 arrays of such nondefective chips could be wired together to produce 256K RAM chips.
+# In order to maximize yield, we want to pack a maximal number of such 2 X 2 arrays into the array of working chips on a wafer."
+
+# To solve this problem, let us start with the necessary imports:
+
+
+using Bloqade
+using GenericTensorNetworks
+using GenericTensorNetworks: unit_disk_graph
+using PythonCall
+using Random
+using Optim
+plt = pyimport("matplotlib.pyplot");
+patches = pyimport("matplotlib.patches");
+np = pyimport("numpy")
+
+# ### Setting up the Problem
+
+# We can visualize this problem as a grid of 64K RAM chips, where some of the grid cells are removed. 
+
+
+fig, ax = plt.subplots(figsize=(5, 5))
+
+G = 5 # grid size (G x G)
+defects = []
+for i in range(1, G*G//15)
+    push!(defects, ((rand(Int)%G+10) % G, (rand(Int)%G + 10) % G))
+end
+
+ax.grid(visible=true)
+plt.xticks(np.arange(1, G+1)) 
+plt.yticks(np.arange(1, G+1)) 
+
+grid = zeros((G, G))
+
+function plot_defects(defects)
+    for (x, y) in defects
+        ax.add_patch(patches.Rectangle((x, y), 1, 1, color="red"))
+        grid[x+1, y+1] = 1
+    end
+end
+
+plot_defects(defects)
+    
+fig
+
+# Wherever there is a 2x2 grid of non-defective chips, these grid cells can be wired together. 
+# This potential grouping of chips can be visualized as a unit disk which overlaps with the four grid cells that are being wired together.
+# For example, in the figure below, the four grid cells in the upper left corner could be wired together, 
+# so we would put a unit disk at (1, 9) with radius 1.
+
+# ![VLSI](../../../assets/vlsi.png)
+
+# The code below shows the unit disk corresponding to every possible wiring.
+# Overlapping unit disks represents two wirings which conflict with each other because they use the same chips. 
+# Our goal is to find the maximum number of non-conflicting wirings, i.e. the MIS.
+
+circs::Vector{Tuple{Int64, Int64}} = []
+Δ = [(0, 0) (0, 1) (1, 0) (1, 1)]
+
+function inrange(x, y)
+    return x>=1 && x<=G && y>=1 && y<=G
+end
+
+for x in 1:G-1
+    for y in 1:G-1
+        works = true
+        for (dx, dy) in Δ
+            if (inrange(x+dx,y+dy))
```

Comment:
I think the following would be more readable?

```suggestion
            if 1 ≤ x+dx ≤ G && 1 ≤ y+dy ≤ G
```